libs/ngrx-toolkit/src/lib/with-mutations.ts

```diff
@@ -0,0 +1,169 @@
+type MutationsDictionary = Record<string, Mutation<any, any>>;
+
+export type MutationResult<T> =
+  | {
+      status: 'success';
+      value: T;
+    }
+  | {
+      status: 'error';
+      error: unknown;
+    }
+  | {
+      status: 'aborted';
+    };
+
+export type MutationStatus = 'idle' | 'processing' | 'error' | 'success';
+
+// withMethods uses Record<string, Function> internally
+// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
+export type MethodsDictionary = Record<string, Function>;
+
+type NamedMutationProps<T extends MutationsDictionary> = {
+  [Prop in keyof T as `${Prop & string}Processing`]: Signal<boolean>;
+} & {
+  [Prop in keyof T as `${Prop & string}Status`]: Signal<MutationStatus>;
+} & {
+  [Prop in keyof T as `${Prop & string}Error`]: Signal<Error | undefined>;
+};
+
+type NamedMutationMethods<T extends MutationsDictionary> = {
+  [Prop in keyof T as `${Prop & string}`]: T[Prop] extends Mutation<
+    infer P,
+    infer R
+  >
+    ? Mutation<P, R>
+    : never;
+};
+
+export type NamedMutationResult<T extends MutationsDictionary> =
+  EmptyFeatureResult & {
+    props: NamedMutationProps<T>;
+    methods: NamedMutationMethods<T>;
+  };
+
+/**
+ * Adds mutation methods to the store. Also, for each mutation method, several
+ * Signals are added informing about the mutation's status and errors.
+ *
+ * ```typescript
+ * export type Params = {
+ *   value: number;
+ * };
+ *
+ * export const CounterStore = signalStore(
+ *   { providedIn: 'root' },
+ *   withState({ counter: 0 }),
+ *   withMutations((store) => ({
+ *     increment: rxMutation({ ... }),
+ *   })),
+ * );
+ * ```
+ *
+ * There are several types of mutations. In the example shown, an {@link module:rx-mutation.rxMutation | rxMutation}
+ * leveraging RxJS is used
+ *
+ * For the defined `increment` mutation, several the following properties and
+ * methods are added to the store:
+ * - `increment(params: Params): Promise<MutationResult<number>>`: The mutation method.
+ * - `incrementProcessing`: A signal indicating if the mutation is in progress.
+ * - `incrementStatus`: A signal representing the current status of the mutation.
+ * - `incrementError`: A signal containing any error that occurred during the mutation.
+ *
+ * @param mutationsFactory
+ */
+export function withMutations<
+  Input extends SignalStoreFeatureResult,
+  Result extends MutationsDictionary,
+>(
+  mutationsFactory: (
+    store: Input['props'] &
+      Input['methods'] &
+      WritableStateSource<Input['state']> &
+      StateSignals<Input['state']>,
+  ) => Result,
+): SignalStoreFeature<Input, NamedMutationResult<Result>>;
+
+export function withMutations<
+  Input extends SignalStoreFeatureResult,
+  Result extends MutationsDictionary,
+>(
+  mutationsFactory: (
+    store: Input['props'] & Input['methods'] & StateSource<Input['state']>,
+  ) => Result,
+): SignalStoreFeature<Input> {
+  return (store) => {
+    // TODO: Is this the correct usage?
```

Comment:
We could leave it is. Might be better options, but we since we are not importing any internal stuff here, it is good to go.